packages/plugin-ext/src/main/browser/view/tree-view-problem-decorator.ts

```diff
@@ -0,0 +1,96 @@
+        super(problemManager);
+    }
+
+    protected collectDecorators(tree: Tree): Map<string, TreeDecoration.Data> {
+        const result = new Map<string, Marker<Diagnostic>>();
+
+        // If the tree root is undefined or the preference for the decorations is disabled, return an empty result map.
+        if (tree.root === undefined || !this.problemPreferences['problems.decorations.enabled']) {
+            return new Map<string, TreeDecoration.Data>();
+        }
+
+        const pathToIdMap = new Map<string, string>();
+
+        for (const node of new DepthFirstTreeIterator(tree.root)) {
+            if (this.isTreeItem(node) && node.resourceUri) {
+                pathToIdMap.set(node.resourceUri.toString(), node.id);
+            }
+        }
+
+        const markers = this.appendContainerItemMarkers(tree, this.collectMarkers(tree), pathToIdMap);
+
+        markers.forEach((marker: Marker<Diagnostic>, uri: string) => {
+            const nodeId = pathToIdMap.get(uri);
+            if (nodeId) {
+                result.set(nodeId, marker);
+            }
+        });
+
+        return new Map(Array.from(result.entries()).map(m => [m[0], this.toDecorator(m[1])]));
+    }
+
+    protected appendContainerItemMarkers(tree: Tree, markers: Marker<Diagnostic>[], pathToIdMap: Map<string, string>): Map<string, Marker<Diagnostic>> {
+        const result: Map<string, Marker<Diagnostic>> = new Map();
+        // We traverse up and assign the diagnostic to the container element.
+        // Just like file based traverse, but use element parent instead.
+        for (const [uri, marker] of new Map<URI, Marker<Diagnostic>>(markers.map(m => [new URI(m.uri), m])).entries()) {
+            const uriString = uri.toString();
+            result.set(uriString, marker);
+            const parentNode = tree.getNode(pathToIdMap.get(uriString))?.parent;
+            if (this.isTreeItem(parentNode) && parentNode.resourceUri) {
+                let parentUri: URI | undefined = new URI(parentNode.resourceUri);
+                while (parentUri && !parentUri.path.isRoot) {
+                    const parentUriString = parentUri.toString();
+                    const existing = result.get(parentUriString);
+                    // Make sure the highest diagnostic severity (smaller number) will be propagated to the container directory.
+                    if (existing === undefined || this.compare(marker, existing) < 0) {
+                        result.set(parentUriString, {
+                            data: marker.data,
+                            uri: parentUriString,
+                            owner: marker.owner,
+                            kind: marker.kind
+                        });
+                        parentUri = parentUri.parent;
+                    } else {
+                        parentUri = undefined;
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
```

Comment:
These two methods contain a lot of redundant iteration and can be simplified considerably. I suggest something like this:
```typescript
protected collectDecorators(tree: Tree): Map<string, TreeDecoration.Data> {
        if (tree.root === undefined || !this.problemPreferences['problems.decorations.enabled']) {
            return new Map();
        }
        const uriToMarker = new Map<string, Marker<Diagnostic>>();
        const markers = this.collectMarkers(tree);
        const idToDecoration = new Map<string, TreeDecoration.Data>();
        const propagateToParents = (uri: URI, childMarkerToPropagate: Marker<Diagnostic>): void => {
            let currentUri = uri.parent;
            while (!currentUri.path.isRoot) {
                const currentUriString = currentUri.toString();
                const existingParentMarker = uriToMarker.get(currentUriString);
                if (!existingParentMarker || this.compare(childMarkerToPropagate, existingParentMarker) < 0) {
                    const parentMarker = {
                        ...childMarkerToPropagate,
                        uri: currentUriString,
                    };
                    uriToMarker.set(currentUriString, parentMarker);
                } else {
                    break; // Higher-priority marker already set and propagated.
                }
                currentUri = currentUri.parent;
            }
        };

        for (const marker of markers) {
            const uri = new URI(marker.uri);
            const uriString = uri.toString();
            uriToMarker.set(uriString, marker);
            propagateToParents(uri, marker);
        }

        for (const node of new DepthFirstTreeIterator(tree.root)) {
            if (this.isTreeItem(node) && node.resourceUri) {
                const marker = uriToMarker.get(node.resourceUri.toString());
                if (marker) {
                    idToDecoration.set(node.id, this.toDecorator(marker));
                }
            }
        }

        return idToDecoration;
    }
```

There's a little bit of confusion in the existing code:
```
// We traverse up and assign the diagnostic to the container element.
// Just like file based traverse, but use element parent instead.
```
First, the code retrieves the parent element (`const parentNode = tree.getNode(pathToIdMap.get(uriString))?.parent;`), but later it uses the URI's parent (`parentUri = parentUri.parent;`). As I've rewritten the code, I follow the latter approach (URI's parent). If the intent is to propagate up through the tree, regardless of whether the tree is structured like a file tree, the code would need to be adjusted accordingly.